File: Cpp/198.cpp
class Solution {
public:

    int f(int i, vector<int>& nums, vector<int>& dp){
        if(i < 0)return 0;
        if(dp[i] != -1)return dp[i];
        return dp[i] = max(nums[i]+f(i-2, nums, dp), max(f(i-2, nums, dp), f(i-1, nums, dp)) );
    }

    int rob(vector<int>& nums) {
        vector<int> dp(nums.size(), -1);
        return max(f(nums.size()-1, nums, dp), f(nums.size()-2, nums, dp));
    }
};

File: Python/198.py
class Solution:
    def f(self, ind, nums, dp):
        if ind < 0:
            return 0
        if dp[ind] != -1:
            return dp[ind]
        # take 
        take = nums[ind] + self.f(ind - 2, nums, dp)
        not_take = self.f(ind-1, nums, dp)
        dp[ind] = max(take, not_take)
        return dp[ind]

    def rob(self, nums: List[int]) -> int:
        dp = [-1]* len(nums)
        return self.f(len(nums)-1, nums, dp)
